src/execution/operator/csv_scanner/sniffer/header_detection.cpp

```diff
@@ -0,0 +1,165 @@
+	}
+
+	// prepend _ if name starts with a digit or is a reserved keyword
+	if (KeywordHelper::IsKeyword(col_name_cleaned) || (col_name_cleaned[0] >= '0' && col_name_cleaned[0] <= '9')) {
+		col_name_cleaned = "_" + col_name_cleaned;
+	}
+	return col_name_cleaned;
+}
+void CSVSniffer::DetectHeader() {
+	// information for header detection
+	bool first_row_consistent = true;
+	// check if header row is all null and/or consistent with detected column data types
+	bool first_row_nulls = true;
+	if (!best_candidate->options.null_padding && best_sql_types_candidates.size() != best_header_row.size()) {
+		// Something went wrong in the auto-detection
+		throw InvalidInputException(
+		    "Error in file \"%s\": CSV options could not be auto-detected. Consider setting parser options manually.",
+		    options.file_path);
+	}
+	for (idx_t col = 0; col < best_header_row.size(); col++) {
+		auto dummy_val = best_header_row[col];
+		if (!dummy_val.IsNull()) {
+			first_row_nulls = false;
+		}
+
+		// try cast to sql_type of column
+		const auto &sql_type = best_sql_types_candidates[col].back();
+		if (!TryCastValue(*best_candidate, dummy_val, sql_type)) {
+			first_row_consistent = false;
+		}
+	}
+
+	// update parser info, and read, generate & set col_names based on previous findings
+	if (((!first_row_consistent || first_row_nulls) && !best_candidate->options.has_header) ||
+	    (best_candidate->options.has_header && best_candidate->options.dialect_options.header)) {
+		best_candidate->dialect_options.header = true;
+		case_insensitive_map_t<idx_t> name_collision_count;
+
+		// get header names from CSV
+		for (idx_t col = 0; col < best_header_row.size(); col++) {
+			const auto &val = best_header_row[col];
+			string col_name = val.ToString();
+
+			// generate name if field is empty
+			if (col_name.empty() || val.IsNull()) {
+				col_name = GenerateColumnName(best_candidate->dialect_options.num_cols, col);
+			}
+
+			// normalize names or at least trim whitespace
+			if (best_candidate->options.normalize_names) {
+				col_name = NormalizeColumnName(col_name);
+			} else {
+				col_name = TrimWhitespace(col_name);
+			}
+
+			// avoid duplicate header names
+			const string col_name_raw = col_name;
+			while (name_collision_count.find(col_name) != name_collision_count.end()) {
+				name_collision_count[col_name] += 1;
+				col_name = col_name + "_" + to_string(name_collision_count[col_name]);
+			}
+
+			names.push_back(col_name);
+			name_collision_count[col_name] = 0;
+		}
+
+	} else {
+		best_candidate->dialect_options.header = false;
+		for (idx_t col = 0; col < best_candidate->dialect_options.num_cols; col++) {
+			string column_name = GenerateColumnName(best_candidate->dialect_options.num_cols, col);
+			names.push_back(column_name);
+		}
+	}
+
+	for (idx_t i = 0; i < MinValue<idx_t>(names.size(), best_candidate->options.name_list.size()); i++) {
+		names[i] = best_candidate->options.name_list[i];
+	}
```

Comment:
What is happening here?

Reply:
It replaces names discovered from the sniffer with names defined by the user. I'm adding a comment.